pubsub/awspubsub/awspubsub.go

```diff
@@ -72,41 +72,71 @@ func openTopic(ctx context.Context, client *sns.SNS, topicARN string) driver.Top
 var stringDataType = aws.String("String")
 
 // SendBatch implements driver.Topic.SendBatch.
-func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) (er error) {
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	// Limit our concurrent RPCs via a semaphore.
+	const maxConcurrency = 10
+	sem := make(chan bool, maxConcurrency)
+	// The result for each message (either nil or an error) will be written to
+	// errc and collected at the end. We'll return the first non-nil error
+	// (if any).
+	errc := make(chan error, len(dms))
 	for _, dm := range dms {
-		attrs := map[string]*sns.MessageAttributeValue{}
-		for k, v := range dm.Metadata {
-			// See the package comments for more details on escaping of metadata
-			// keys & values.
-			k = escape.HexEscape(k, func(runes []rune, i int) bool {
-				c := runes[i]
-				switch {
-				case escape.IsASCIIAlphanumeric(c):
-					return false
-				case c == '_' || c == '-':
-					return false
-				case c == '.' && i != 0 && runes[i-1] != '.':
-					return false
-				}
-				return true
-			})
-			attrs[k] = &sns.MessageAttributeValue{
-				DataType:    stringDataType,
-				StringValue: aws.String(escape.URLEscape(v)),
-			}
+		select {
+		case <-ctx.Done():
+			errc <- ctx.Err()
+			break
+		case sem <- true:
 		}
-		_, err := t.client.Publish(&sns.PublishInput{
-			Message:           aws.String(string(dm.Body)),
-			MessageAttributes: attrs,
-			TopicArn:          &t.arn,
-		})
+		go func(dm *driver.Message) {
+			errc <- t.sendMessage(ctx, dm)
```

Comment:
Important question: do you want to stop the other sendMessages on the first error you see? If so, that would probably tip the balance to using `errgroup`.

Reply:
I don't think so.